R/bitbucket.R

```diff
@@ -0,0 +1,138 @@
+# Copy of code from install_github.R.
+# release tag is not supported by Bitbucket
+parse_bitbucket_repo <- function(path) {
+  username_rx <- "(?:([^/]+)/)?"
+  repo_rx <- "([^/@#]+)"
+  subdir_rx <- "(?:/([^@#]*[^@#/]))?"
+  ref_rx <- "(?:@([^*].*))"
+  pull_rx <- "(?:#([0-9]+))"
+  ref_or_pull_rx <- sprintf("(?:%s|%s)?", ref_rx, pull_rx)
+  bitbucket_rx <- sprintf("^(?:%s%s%s%s|(.*))$",
+    username_rx, repo_rx, subdir_rx, ref_or_pull_rx)
+
+  param_names <- c("username", "repo", "subdir", "ref", "pull", "invalid")
+  replace <- stats::setNames(sprintf("\\%d", seq_along(param_names)), param_names)
+  params <- lapply(replace, function(r) gsub(bitbucket_rx, r, path, perl = TRUE))
+  if (params$invalid != "")
+    stop(sprintf("Invalid Bitbucket repo: %s", path))
+  params <- params[sapply(params, nchar) > 0]
+
+  if (!is.null(params$pull)) {
+    params$ref <- bitbucket_pull(params$pull)
+    params$pull <- NULL
+  }
+
+  params
+}
+
+#' Bitbucket references
+#'
+#' Use as \code{ref} parameter to \code{\link{install_bitbucket}}.
+#' Allows installing a specific pull request.
+#'
+#' @param pull The pull request to install
+#' @seealso \code{\link{install_bitbucket}}
+#' @export
+bitbucket_pull <- function(pull) structure(pull, class = "bitbucket_pull")
+
+bitbucket_resolve_ref <- function(x, params) UseMethod("bitbucket_resolve_ref")
+
+#' @export
+bitbucket_resolve_ref.default <- function(x, params) {
+  params$ref <- x
+  params
+}
+
+#' @export
+bitbucket_resolve_ref.NULL <- function(x, params) {
+  params$ref <- "master"
+  params
+}
+
+#' @export
+bitbucket_resolve_ref.bitbucket_pull <- function(x, params, ..., api_version) {
+  # GET /repositories/{owner}/{repo_slug}/pullrequests/{id}
+  # https://confluence.atlassian.com/bitbucket/pullrequests-resource-423626332.html#pullrequestsResource-GETaspecificpullrequest
+  path <- file.path("repositories", params$username, params$repo,
+    "pullrequests", x)
+  response <- bitbucket_GET(path, ..., host = params$host,
+    api_version = api_version)
+
+  params$username <- response$author$username
+  params$ref <- response$source$branch$name
+  params
+}
+
+bitbucket_host <- function (host = NULL) {
```

Comment:
This seems an unnecessary complication, why can't you just have `https://api.bitbucket.org` be the default host?